### PR TITLE
aws.rds - type: db-parameter fix cache usage

### LINKED ISSUE #7791: sqlkvcache errors: AttributeError: __enter__ and sqlite3.InterfaceError: Error binding parameter 1 - probably unsupported type.
### Describe the bug

```AttributeError: __enter__```

### What did you expect to happen?

I expected 2 + 2 to equal 4, but instead 2 + 2 equaled 6!

### Cloud Provider

Amazon Web Services (AWS)

### Cloud Custodian version and dependency information

_No response_

### Policy

```shell
- name: rds-invalid-param-unmark
  resource: aws.rds
  filters:
    - type: db-parameter
```

### Relevant log/traceback output

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/c7n/policy.py", line 302, in run
    resources = self.policy.resource_manager.resources()
  File "/usr/local/lib/python3.7/dist-packages/c7n/query.py", line 535, in resources
    resources = self.filter_resources(resources)
  File "/usr/local/lib/python3.7/dist-packages/c7n/manager.py", line 111, in filter_resources
    resources = f.process(resources, event)
  File "/usr/local/lib/python3.7/dist-packages/c7n/filters/core.py", line 378, in process
    return self.process_set(resources, event)
  File "/usr/local/lib/python3.7/dist-packages/c7n/filters/core.py", line 397, in process_set
    resources = f.process(resources, event)
  File "/usr/local/lib/python3.7/dist-packages/c7n/filters/core.py", line 337, in process
    return self.process_set(resources, event)
  File "/usr/local/lib/python3.7/dist-packages/c7n/filters/core.py", line 353, in process_set
    r[rtype_id] for r in f.process(resources, event)])
  File "/usr/local/lib/python3.7/dist-packages/c7n/filters/core.py", line 364, in process
    resources = f.process(resources, events)
  File "/usr/local/lib/python3.7/dist-packages/c7n/filters/core.py", line 337, in process
    return self.process_set(resources, event)
  File "/usr/local/lib/python3.7/dist-packages/c7n/filters/core.py", line 353, in process_set
    r[rtype_id] for r in f.process(resources, event)])
  File "/usr/local/lib/python3.7/dist-packages/c7n/resources/rds.py", line 1650, in process
    pg_values = self.manager._cache.get(cache_key)
  File "/usr/local/lib/python3.7/dist-packages/c7n/cache.py", line 143, in get
    with self.conn as cursor:
AttributeError: __enter__
```

### Extra information or context

Another policy that ran for more than a day re: ecs-task-definition

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/dist-packages/c7n/commands.py", line 302, in run
    policy()
  File "/usr/local/lib/python3.7/dist-packages/c7n/policy.py", line 1296, in __call__
    resources = mode.run()
  File "/usr/local/lib/python3.7/dist-packages/c7n/policy.py", line 302, in run
    resources = self.policy.resource_manager.resources()
  File "/usr/local/lib/python3.7/dist-packages/c7n/query.py", line 531, in resources
    self._cache.save(cache_key, resources)
  File "/usr/local/lib/python3.7/dist-packages/c7n/cache.py", line 162, in save
    (sqlite3.Binary(encode(key)), sqlite3.Binary(encode(data)), timestamp))
sqlite3.InterfaceError: Error binding parameter 1 - probably unsupported type.
```

### PR DESCRIPTION
**Pull request by PratMis**
Closes #7791 
This fixes the error wrt rds policies with filter `type: db-parameter`. The other error feels like a problem on our end on one of our legacy accounts where the number of resources are very high
Also, added a test case for codecov

### CODE REVIEW

**Comment by PratMis** [comment]
@kapilt / @thisisshi when you have some time, do you mind taking a look please? Thanks

File: c7n/resources/rds.py [hunk 1]
```diff
@@ -1632,31 +1632,35 @@ def recast(val, datatype):
 
         return ret_val
 
+    def handle_paramgroup_cache(self, client, paginator, param_groups):
+        pgcache = {}
+        cache = self.manager._cache
+
+        for pg in param_groups:
+            with cache:
```
Comment (kapilt) [inline c7n/resources/rds.py:1640]: move the context manager outside of the loop

**Review by kapilt — APPROVED** [review]
lgtm

**Comment by PratMis** [comment]
> if you do see the other cache issue again re invalid type, please do file another issue, I'm curious to track that one down, it does some unrelated to this fix though.

Definitely, i'll keep an eye on our run logs

**Pull request merged**